### PR TITLE
Skip orchestrator deployment if the live image matches the latest one.

### PR DESCRIPTION
**Pull request by n-g**

### CODE REVIEW

File: orchestration/prepare.go [hunk 1]
```diff
@@ -47,33 +75,110 @@ func PrepareOrchestrator(ctx context.Context, targetEnv cfg.Configuration, clust
...
+		res.Pinned = append(res.Pinned, &schema.Workspace_BinaryDigest{
+			PackageName: prebuilt.PackageName,
+			Repository:  prebuilt.Repository,
+			Digest:      prebuilt.Digest,
+		})
+	}
+
+	return res, nil
+}
+func getVersionsFromOrchestrator(ctx context.Context, targetEnv cfg.Configuration, cluster runtime.Cluster) (*proto.GetOrchestratorVersionResponse, error) {
```
Comment (hugosantos) [inline orchestration/prepare.go:156]: nit: newline

File: orchestration/prepare.go [hunk 2]
```diff
@@ -47,33 +75,110 @@ func PrepareOrchestrator(ctx context.Context, targetEnv cfg.Configuration, clust
...
+		return nil, err
+	}
+
+	// Only dial once.
+	conn, err := boundCluster.DialServer(ctx, server, &schema.Endpoint_Port{Name: portName})
+	if err != nil {
+		return nil, err
+	}
+
+	rpc, err := grpc.DialContext(ctx, "orchestrator",
```
Comment (hugosantos) [inline orchestration/prepare.go:173]: whats the timeout here?

File: orchestration/service/versions.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	"k8s.io/client-go/rest"
+	"namespacelabs.dev/foundation/framework/kubernetes/kubedef"
+	"namespacelabs.dev/foundation/internal/artifacts/oci"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/orchestration/proto"
+	"namespacelabs.dev/foundation/schema"
+)
+
+const (
+	serverId                      = "0fomj22adbua2u0ug3og"
```
Comment (hugosantos) [inline orchestration/service/versions.go:26]: move to shared constants package?

File: orchestration/service/versions.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	// shedule a non-blocking update so that future calls get the latest version.
+	go vc.updateLatest()
+
+	return &proto.GetOrchestratorVersionResponse{
+		Current: vc.current,
+		Pinned:  vc.pinned,
+	}
+}
+
+func getCurrentVersion(ctx context.Context) (*schema.Workspace_BinaryDigest, error) {
```
Comment (hugosantos) [inline orchestration/service/versions.go:81]: Why not use RuntimeConfig.Current.ImageRef?

File: orchestration/service/versions.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+	return vc
+}
+
+func (vc *versionChecker) GetOrchestratorVersion() *proto.GetOrchestratorVersionResponse {
+	vc.mu.Lock()
+	defer vc.mu.Unlock()
+
+	// shedule a non-blocking update so that future calls get the latest version.
+	go vc.updateLatest()
```
Comment (hugosantos) [inline orchestration/service/versions.go:73]: I wouldn't do this -- I would add a flag "skip_cache" that instructs the orchestrator to ignore its cache and fetch the latest version.

File: orchestration/prepare.go [hunk 6]
```diff
@@ -6,22 +6,45 @@ package orchestration
...
+	serverName                              = "orchestration-api-server"
+	serverPkg            schema.PackageName = "namespacelabs.dev/foundation/orchestration/server"
+	toolPkg              schema.PackageName = "namespacelabs.dev/foundation/orchestration/server/tool"
+)
 
 var (
 	UseOrchestrator              = true
+	UsePinnedOrchestrator        = true
 	RenderOrchestratorDeployment = false
+	ForceOrchestratorDeployment  = false
```
Comment (hugosantos) [inline orchestration/prepare.go:40]: Am getting a bit lost in all of these permutations. :-D

useOrchestrator seems important
lets collapse usepinned and force? e.g. "--use_head_orchestrator", which both skips version check and unconditionally deploys the latest built on.

File: orchestration/service/versions.go (hunk 4 above)
Comment (n-g) [reply orchestration/service/versions.go:81]: 🤦

**Pull request merged**